addons/godot-next/nodes/behaviors.gd

```diff
@@ -0,0 +1,178 @@
+# author: xdgamestudios
+# license: MIT
+# description: Provides a behavior system API for the nodes. 
+# todo: Refactor code in general to clean and improve performance
+# usage:
+# - Creating:
+#     behaviors = Behaviors.new(); # through code
+#     Note:
+#     - To use in the inspector just add node of type Behaviors
+# - Adding Behaviors:
+#     behaviors.add_behavior(MyBehavior) # returns instance of MyBehavior
+#     Note:
+#     - Only one behavior per type is allowed.
+#     - If invalid behavior class is provided returns [null].
+#     - If behavior already exists current instance is returned.
+# - Checking Behaviors:
+#     behaviors.has_behavior(MyBehavior) # returns true if behavior exists within the collection
+# - Retrieving Behaviors:
+#     behaviors.get_behavior(MyBehavior) # return the instance of the behavior type
+#     Note:
+#     - Returns [null] if behavior doesn't exist in the collection
+# - Removing Behaviors:
+#     behaviors.remove_behavior(MyBehavior) # removes the behavior from the current collection
+#     Note:
+#     - Returns [true] if the action was successful else return [false]
+tool
+extends Node
+class_name Behaviors
+
+##### PROPERTIES #####
+
+var _behaviors: ResourceSet
+
+var _callbacks: Dictionary = {
+	"_ready" : {},
+	"_process" : {},
+	"_physics_process" : {},
+	"_input" : {},
+	"_unhandled_input" : {},
+	"_unhandled_key_input" : {}
+}
+
+##### NOTIFICATIONS #####
+
+func _set(p_property: String, p_value) -> bool:
+	if _behaviors.can_contain(p_property):
+		#warning-ignore:return_value_discarded
+		_behaviors.set_element(p_property, p_value)
+		property_list_changed_notify()
+		return true
+	return false
+
+func _get(p_property: String):
+	if _behaviors.can_contain(p_property):
+		return _behaviors.get_element(p_property)
+	return null
+
+func _get_property_list() -> Array:
+	if not _behaviors:
+		_behaviors = ResourceSet.new()
+	_behaviors.setup("_behaviors", Behavior)
+	var list = [];
+	list += [ PropertyInfo.new_dictionary("_behaviors", 0, "", PROPERTY_USAGE_STORAGE).to_dict() ];
+	list += _behaviors.get_collection_property_list()
+	return list
+
+func _ready() -> void:
+	if not Engine.editor_hint:
+		var behaviors = _behaviors.get_data();
+		for behavior in behaviors.values():
+			_initialize_behavior(behavior)
+		
+		for behavior in _callbacks["_ready"]:
+			behavior._ready()
+
+func _process(delta: float) -> void:
```

Comment:
For the following 5 methods, could you call the appropriate `set_process_<notification>(enable)` method as the behavior script changes and you update the callbacks? If there are no Behaviors in the relevant `_callbacks` list, or if there are behaviors, but none of them are enabled, then we should toggle off the method call completely so we don't waste time executing the scripting API callback and iterating over the behaviors.

Reply:
if there are no behaviors with that methods then there will be no iterations as they will not be in the list. Not sure I understood the rest!

Reply:
I was referring to the Behaviors class's callbacks (which do the looping operations). When you update, the callbacks, you can determine whether any Behavior resources would execute a callback (if they have it implemented and are enabled).

    1. When someone toggles whether a Resource is enabled, have that *also* update whether that resource is registered in the `_callbacks` sub-array.
    2. If 1 is done, there is no longer any need to check for "enabled" in the loops because it will only be present in the `_callbacks` sub-array if it *is* enabled.
    3. When updating the callbacks in the Behaviors class, you can check if any given sub-array of `_callbacks` is empty. If it is, then you can completely toggle off Behaviors' calling of that notification entirely. For example, if no Behavior resources are present with the `_process` callback, then you can do `set_process(false)` to prevent the engine from calling `_process` on the Behaviors node.